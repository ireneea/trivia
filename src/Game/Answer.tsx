import React from "react";
import { View, Text } from "react-native";

const Answer = ({ answer }) => {
  return (
    <View>
      <Text>{answer?.answer}</Text>
    </View>
  );
};

export default Answer;
