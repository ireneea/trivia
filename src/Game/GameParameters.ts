export const CORRECT_ANSWER_POINTS = 100;
export const READ_ANSWER_TIME = 1000;
export const ANSWER_TIME = 5000;
export const ANSWER_BONUS_TIME_LIMIT = 3000;
