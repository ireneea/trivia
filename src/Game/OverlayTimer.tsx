import React from "react";
import { View, StyleSheet } from "react-native";
import _isNumber from "lodash/isNumber";

type Props = {
  timeLeft?: number;
  totalTime?: number;
};

const OverlayTimer: React.FC<Props> = (props) => {
  const height = { height: `${getTimeSpentPercentage(props)}%` };

  return <View style={StyleSheet.flatten([styles.overlay, height])} testID="overlay"></View>;
};

const getTimeSpentPercentage = (props: Props) => {
  let percent = 0;
  const { timeLeft, totalTime } = props;
  if (_isNumber(timeLeft) && _isNumber(totalTime) && totalTime > 0 && timeLeft >= 0) {
    if (timeLeft > totalTime) {
      percent = 0;
    } else {
      percent = Math.round(100 - (timeLeft * 100) / totalTime);
    }
  }

  return percent;
};

const styles = StyleSheet.create({
  overlay: {
    flex: 1,
    position: "absolute",
    left: 0,
    top: 0,
    opacity: 0.05,
    backgroundColor: "black",
    width: "100%",
    height: "0%",
  },
});

export default OverlayTimer;
