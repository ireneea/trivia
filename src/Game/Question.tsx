import React from "react";
import { View, Text } from "react-native";

const Question = ({ question }) => {
  return (
    <View testID="question">
      <Text>{question?.text}</Text>
    </View>
  );
};

export default Question;
