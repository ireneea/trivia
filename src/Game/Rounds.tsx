import React from "react";
import { View } from "react-native";
import { FontAwesome } from "@expo/vector-icons";

import { AnswerResult, GameResults } from "../ts/appTypes";

type Prop = {
  rounds: number;
  currentRound: number;
  results: GameResults;
};

enum RoundType {
  CURRENT = "CURRENT",
  SUCCESS = "SUCCESS",
  FAIL = "FAIL",
  INACTIVE = "INACTIVE",
}

const Rounds: React.FC<Prop> = (props) => {
  const { rounds, currentRound, results } = props;
  const indicators = [];
  for (let i = 1; i <= rounds; i++) {
    const result = results[`${i}`];
    let type: RoundType = RoundType.INACTIVE;

    if (i > currentRound) {
      type = RoundType.INACTIVE;
    } else if (result === AnswerResult.CORRECT) {
      type = RoundType.SUCCESS;
    } else if (result === AnswerResult.INCORRECT || result === AnswerResult.NO_ANSWER) {
      type = RoundType.FAIL;
    } else {
      type = RoundType.CURRENT;
    }

    let indicatorsProps = {
      active: i <= currentRound,
      key: i,
      type,
    };

    indicators.push(indicatorsProps);
  }

  return (
    <View style={{ flex: 1, flexDirection: "row", justifyContent: "space-between", height: 15 }}>
      {indicators.map(({ active, key, type }) => {
        let res = null;

        if (type === RoundType.INACTIVE) {
          res = <Inactive key={key} />;
        } else if (type === RoundType.SUCCESS) {
          res = <Success key={key} />;
        } else if (type === RoundType.FAIL) {
          res = <Fail key={key} />;
        } else {
          res = <Current key={key} />;
        }

        return res;
      })}
    </View>
  );
};

const Current = () => <FontAwesome testID="current" name="circle" size={12} color="rgba(0, 0, 0, 0.1)" />;
const Inactive = () => <FontAwesome testID="inactive" name="circle-o" size={12} color="rgba(0, 0, 0, 0.1)" />;
const Success = () => <FontAwesome testID="success" name="circle" size={12} color="green" />;
const Fail = () => <FontAwesome testID="fail" name="times" size={12} color="red" />;

export default Rounds;
