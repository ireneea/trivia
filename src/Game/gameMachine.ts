import { Machine, MachineConfig, assign } from "xstate";

export const events = {
  START: "START",
  ANSWER: "ANSWER",
  CORRECT_ANSWER: "CORRECT_ANSWER",
  INCORRECT_ANSWER: "INCORRECT_ANSWER",
  NO_ANSWER: "NO_ANSWER",
  NEXT_ROUND: "NEXT_ROUND",
  QUIT: "QUIT",
};

const config: MachineConfig<any, any, any> = {
  id: "game",
  initial: "idling",
  context: {
    rounds: 0,
    currentRound: 0,
    score: 0,
    results: {},
  },
  on: {
    [events.QUIT]: ".gameOver",
  },
  states: {
    idling: {
      on: {
        [events.START]: {
          target: "answering",
          cond: "isGameValid",
        },
      },
    },
    answering: {
      entry: "onQuestionStart",
      on: {
        [events.CORRECT_ANSWER]: "feedback.correct",
        [events.INCORRECT_ANSWER]: "feedback.incorrect",
        [events.NO_ANSWER]: "feedback.noAnswer",
      },
    },
    feedback: {
      on: {
        [events.NEXT_ROUND]: [
          {
            target: "answering",
            cond: "isNotLastRound",
          },
          {
            target: "gameOver",
            cond: "isLastRound",
          },
        ],
      },
      states: {
        correct: {
          entry: "onCorrectAnswer",
        },
        incorrect: {
          entry: "onIncorrectAnswer",
        },
        noAnswer: {
          entry: "onNoAnswer",
        },
      },
    },
    gameOver: {
      type: "final",
    },
  },
};

const guards = {
  isGameValid: (ctx) => {
    const isValid = !!ctx.rounds && ctx.rounds > 0;
    return isValid;
  },
  isNotLastRound: (ctx) => {
    const { currentRound, rounds } = ctx;
    return currentRound < rounds;
  },
  isLastRound: (ctx) => {
    const { currentRound, rounds } = ctx;
    return currentRound === rounds;
  },
};

const actions = {
  onQuestionStart: assign({
    currentRound: (ctx) => ctx.currentRound + 1,
  }),
  onCorrectAnswer: assign({
    score: (ctx, event) => {
      const points = event?.points || 0;
      return ctx.score + points;
    },
    results: (ctx, event) => {
      return {
        ...ctx.results,
        [ctx.currentRound]: "correct",
      };
    },
  }),
  onIncorrectAnswer: assign({
    results: (ctx, event) => {
      return {
        ...ctx.results,
        [ctx.currentRound]: "incorrect",
      };
    },
  }),
  onNoAnswer: assign({
    results: (ctx, event) => {
      return {
        ...ctx.results,
        [ctx.currentRound]: "noAnswer",
      };
    },
  }),
};

export default {
  events,
  config,
  options: {
    guards,
    actions,
  },
  machine: Machine(config, {
    guards,
    actions,
  }),
};
