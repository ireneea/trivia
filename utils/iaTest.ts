import { ReactTestInstance } from "react-test-renderer";

export function findByTestID(
  root: ReactTestInstance,
  testID: string
): ReactTestInstance {
  return root.findByProps({ testID });
}

export function mockNavigationProps(props = {}) {
  return {
    navigation: {
      // when use in a unit test jest will be defined
      navigate: jest.fn(),
    },
    ...props,
  };
}
