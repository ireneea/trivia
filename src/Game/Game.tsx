import React, { useState, useEffect } from "react";
import { View, Text, TouchableWithoutFeedback } from "react-native";
import { globalStyles } from "../styles/globals";

import Answer from "./Answer";
import Question from "./Question";

type Props = {
  navigation: any;
  route: any;
};

type Game = {
  questions: Question[];
};

type Question = {
  text: string;
  choices: Answer[];
};

type Answer = {
  answer: string;
};

const Game: React.FC<Props> = (props) => {
  const game: Game = props?.route?.params?.game;

  const [currentQuestion, setCurrentQuestion] = useState(():
    | Question
    | undefined => {
    return undefined;
  });

  useEffect(() => {
    let firstQuestion;
    if (game && Array.isArray(game.questions)) {
      firstQuestion = game.questions[0];
    }

    setCurrentQuestion(firstQuestion);
  }, [game]);

  const onEndGame = () => {
    const { navigation } = props;
    navigation.navigate("Score");
  };

  return (
    <View style={globalStyles.centredContainer}>
      <TouchableWithoutFeedback testID="endGameBtn" onPress={onEndGame}>
        <Text>End Game</Text>
      </TouchableWithoutFeedback>
      <Question question={currentQuestion} />
      {currentQuestion?.choices.map((answer) => (
        <Answer key={answer.answer} answer={answer} />
      ))}
    </View>
  );
};

export default Game;
