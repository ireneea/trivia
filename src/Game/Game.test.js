import React from "react";
import { render, fireEvent } from "@testing-library/react-native";

import Game from "./Game";

import { mockNavigationProps, mockRouteParamsProps } from "../../utils/iaTest";

const game = {
  questions: [
    {
      text: "Kampala is the capital of",
      correctAnswer: "two",
      choices: [
        { answer: "one" },
        { answer: "two" },
        { answer: "three" },
        { answer: "four" },
      ],
    },
  ],
};

describe("Game", () => {
  it("render Game correctly", () => {
    const { baseElement } = render(<Game />);
    expect(baseElement).toMatchSnapshot();
  });

  it("navigate to Score", () => {
    const props = mockNavigationProps();

    const { getByText } = render(<Game {...props} />);
    fireEvent.press(getByText("End Game"));

    expect(props.navigation.navigate).toHaveBeenCalledWith("Score");
  });

  it("render first question", () => {
    const { getByText } = render(<Game {...mockRouteParamsProps({ game })} />);
    const [firstQuestion] = game.questions;
    expect(getByText(firstQuestion.text)).toBeTruthy();
  });

  it("render answers", () => {
    const { getByText } = render(<Game {...mockRouteParamsProps({ game })} />);
    const [firstQuestion] = game.questions;
    const { choices } = firstQuestion;

    expect(getByText(choices[0].answer)).toBeTruthy();
    expect(getByText(choices[1].answer)).toBeTruthy();
    expect(getByText(choices[2].answer)).toBeTruthy();
    expect(getByText(choices[3].answer)).toBeTruthy();
  });
});
