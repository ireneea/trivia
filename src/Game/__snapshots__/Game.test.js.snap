// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game render Game correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TouchableWithoutFeedback
        testID="endGameBtn"
      >
        <Text>
          End Game
        </Text>
      </TouchableWithoutFeedback>
      <View
        testID="question"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
